File: dotnet/src/HybridRow.Tests.Unit/PermuteExtensions.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Tests.Unit
{
    using System.Collections.Generic;
    using System.Linq;

    /// <summary>
    /// Extension methods for computing permutations of <see cref="IEnumerable{T}"/>.
    /// </summary>
    public static class PermuteExtensions
    {
        /// <summary>Generate all permutations of a given enumerable.</summary>
        public static IEnumerable<IEnumerable<T>> Permute<T>(this IEnumerable<T> list)
        {
            int start = 0;
            foreach (T element in list)
            {
                int index = start;
                T[] first = { element };
                IEnumerable<T> rest = list.Where((s, i) => i != index);
                if (!rest.Any())
                {
                    yield return first;
                }

                foreach (IEnumerable<T> sub in rest.Permute())
                {
                    yield return first.Concat(sub);
                }

                start++;
            }
        }
    }
}

File: jre/src/test/java/com/azure/data/cosmos/serialization/hybridrow/unit/PermuteExtensions.java
//------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
//------------------------------------------------------------

package com.azure.data.cosmos.serialization.hybridrow.unit;

/**
 * Extension methods for computing permutations of {@link IEnumerable{T}}.
 */
public final class PermuteExtensions {
    /**
     * Generate all permutations of a given enumerable.
     */
    public static <T> java.lang.Iterable<java.lang.Iterable<T>> Permute(java.lang.Iterable<T> list) {
        int start = 0;
        for (T element : list) {
            int index = start;
            T[] first = { element };
            java.lang.Iterable<T> rest = list.Where((s, i) -> i != index);
            if (!rest.Any()) {
                // TODO: C# TO JAVA CONVERTER: Java does not have an equivalent to the C# 'yield' keyword:
                yield return first;
            }

            for (java.lang.Iterable<T> sub :
                PermuteExtensions.Permute(rest)) {
                // TODO: C# TO JAVA CONVERTER: Java does not have an equivalent to the C# 'yield' keyword:
                yield return first.Concat(sub);
            }

            start++;
        }
    }
}